loopchain/blockchain/epoch.py

```diff
@@ -14,40 +14,50 @@
 """It manages the information needed during consensus to store one block height.
 Candidate Blocks, Quorum, Votes and Leader Complaints.
 """
-import loopchain.utils as util
+import logging
+import traceback
 
-from loopchain.baseservice import ObjectManager
-from loopchain.blockchain import Vote
+import loopchain.utils as util
 from loopchain import configure as conf
+from loopchain.baseservice import ObjectManager
+from loopchain.blockchain import Vote, BlockBuilder, Transaction, TransactionStatusInQueue, TransactionVerifier
 
 
 class Epoch:
     COMPLAIN_VOTE_HASH = "complain_vote_hash_for_reuse_Vote_class"
 
-    def __init__(self, height: int, leader_id=None):
-        util.logger.notice(f"New Epoch Start height({height}) leader_id({leader_id})")
-        self.height = height
+    def __init__(self, block_manager, leader_id=None):
+        if block_manager.get_blockchain().last_block:
+            self.height = block_manager.get_blockchain().last_block.header.height + 1
+        else:
+            self.height = 1
         self.leader_id = leader_id
+        self.prev_leader_id = None
+        self.__block_manager = block_manager
+        self.__blockchain = self.__block_manager.get_blockchain()
+        util.logger.debug(f"New Epoch Start height({self.height }) leader_id({leader_id})")
 
         # TODO using Epoch in BlockManager instead using candidate_blocks directly.
         # But now! only collect leader complain votes.
         self.__candidate_blocks = None
         self.__complain_vote = Vote(Epoch.COMPLAIN_VOTE_HASH, ObjectManager().channel_service.peer_manager)
```

Comment:
Why does it use dummy hash?

Reply:
Every epoch has only one complain vote object.